core/chains/terra/terratxm/txm.go

```diff
@@ -0,0 +1,400 @@
+	// There appears to be no gas limit per tx, only per block, so theoretically
+	// we could include 1000 msgs which use up to 1M gas.
+	// To be conservative and since the number of messages we'd
+	// have in a batch on average roughly correponds to the number of terra ocr jobs we're running (do not expect more than 100),
+	// we can set a max msgs per batch of 100.
+	MaxMsgsPerBatch = 100
+
+	// BlocksUntilTxTimeout ~8s per block, so ~80s until we give up on the tx getting confirmed
+	// Anecdotally it appears anything more than 4 blocks would be an extremely long wait.
+	BlocksUntilTxTimeout = 10
+)
+
+// Txm manages transactions for the terra blockchain.
+type Txm struct {
+	starter           utils.StartStopOnce
+	eb                pg.EventBroadcaster
+	sub               pg.Subscription
+	ticker            *time.Ticker
+	orm               *ORM
+	lggr              logger.Logger
+	tc                terraclient.ReaderWriter
+	ks                keystore.Terra
+	stop, done        chan struct{}
+	confirmPollPeriod time.Duration
+	confirmMaxPolls   int
+}
+
+// NewTxm creates a txm
+func NewTxm(db *sqlx.DB, tc terraclient.ReaderWriter, ks keystore.Terra, lggr logger.Logger, cfg pg.LogConfig, eb pg.EventBroadcaster, pollPeriod time.Duration) *Txm {
+	ticker := time.NewTicker(pollPeriod)
+	return &Txm{
+		starter:           utils.StartStopOnce{},
+		eb:                eb,
+		orm:               NewORM(db, lggr, cfg),
+		ks:                ks,
+		ticker:            ticker,
+		tc:                tc,
+		lggr:              lggr,
+		stop:              make(chan struct{}),
+		done:              make(chan struct{}),
+		confirmPollPeriod: 1 * time.Second,
+		confirmMaxPolls:   100,
+	}
+}
+
+// Start subscribes to pg notifications about terra msg inserts and processes them.
+func (txm *Txm) Start() error {
+	return txm.starter.StartOnce("terratxm", func() error {
+		sub, err := txm.eb.Subscribe(pg.ChannelInsertOnTerraMsg, "")
+		if err != nil {
+			return err
+		}
+		txm.sub = sub
+		go txm.run(sub)
+		return nil
+	})
+}
+
+func (txm *Txm) confirmAnyUnconfirmed() {
+	// Confirm any broadcasted but not confirmed txes.
+	// This is an edge case if we crash after having broadcasted but before we confirm.
+	broadcasted, err := txm.orm.SelectMsgsWithState(Broadcasted)
+	if err != nil {
+		// Should never happen but if so, theoretically can retry with a reboot
+		txm.lggr.Errorw("unable to look for broadcasted but unconfirmed txes", "err", err)
+		return
+	}
+	if len(broadcasted) == 0 {
+		return
+	}
+	msgsByTxHash := make(map[string][]TerraMsg)
+	for _, msg := range broadcasted {
+		msgsByTxHash[*msg.TxHash] = append(msgsByTxHash[*msg.TxHash], msg)
+	}
+	for txHash, msgs := range msgsByTxHash {
+		err := txm.confirmTx(txHash, getIDs(msgs))
+		if err != nil {
+			txm.lggr.Errorw("unable to confirm broadcasted but unconfirmed txes", "err", err)
+		}
+	}
+}
+
+func (txm *Txm) run(sub pg.Subscription) {
+	defer func() { txm.done <- struct{}{} }()
```

Comment:
Could close instead for readability?
```suggestion
	defer close(txm.done)
```